File: 07-Builder/src/java/chapter07/House.java
package chapter07;

public class House {
    private final int rooms;
    private final int floors;
    private final boolean hasGarage;
    private final boolean hasPool;
    private final boolean hasGarden;
    private final String roofType;

    // Private constructor — only Builder can create a House
    House(int rooms, int floors, boolean hasGarage, boolean hasPool,
          boolean hasGarden, String roofType) {
        this.rooms = rooms;
        this.floors = floors;
        this.hasGarage = hasGarage;
        this.hasPool = hasPool;
        this.hasGarden = hasGarden;
        this.roofType = roofType;
    }

    public int getRooms() { return rooms; }
    public int getFloors() { return floors; }
    public boolean hasGarage() { return hasGarage; }
    public boolean hasPool() { return hasPool; }
    public boolean hasGarden() { return hasGarden; }
    public String getRoofType() { return roofType; }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        sb.append("House [")
          .append(rooms).append(" rooms, ")
          .append(floors).append(" floors, ")
          .append("roof=").append(roofType);
        if (hasGarage) sb.append(", garage");
        if (hasPool) sb.append(", pool");
        if (hasGarden) sb.append(", garden");
        sb.append("]");
        return sb.toString();
    }
}

File: 07-Builder/src/go/house.go
package main

import "fmt"

type House struct {
	rooms     int
	floors    int
	hasGarage bool
	hasPool   bool
	hasGarden bool
	roofType  string
}

func (h House) String() string {
	s := fmt.Sprintf("House [%d rooms, %d floors, roof=%s", h.rooms, h.floors, h.roofType)
	if h.hasGarage {
		s += ", garage"
	}
	if h.hasPool {
		s += ", pool"
	}
	if h.hasGarden {
		s += ", garden"
	}
	s += "]"
	return s
}
